components/supervisor/pkg/supervisor/top.go

```diff
@@ -10,17 +10,68 @@ import (
 	"os"
 	"strconv"
 	"strings"
+	"sync"
 	"time"
 
 	"golang.org/x/xerrors"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/credentials/insecure"
 
 	linuxproc "github.com/c9s/goprocinfo/linux"
+	"github.com/gitpod-io/gitpod/common-go/log"
 	"github.com/gitpod-io/gitpod/supervisor/api"
 	daemonapi "github.com/gitpod-io/gitpod/ws-daemon/api"
 )
 
+type TopService struct {
+	data      *api.ResourcesStatusResponse
+	ready     chan struct{}
+	readyOnce sync.Once
+}
+
+func (t *TopService) GetTopData() *api.ResourcesStatusResponse {
+	var newData *api.ResourcesStatusResponse = t.data
+	return newData
+}
+
+// Observe starts observing the resource status
+func (t *TopService) Observe(ctx context.Context) {
+	var (
+		delay                       = 1 * time.Second
+		reconnectionDelayGrowFactor float64
+		minReconnectionDelay        = 1 * time.Second
+		maxReconnectionDelay        = 30 * time.Second
+	)
+
+	t.ready = make(chan struct{})
+
+	go func() {
+		for {
+			data, err := Top(ctx)
+			if err != nil {
+				log.WithField("error", err).Error("Failed to retrieve resource status from upstream")
```

Comment:
usually we will add something like `trying again in $d...` see [[1](https://github.com/gitpod-io/gitpod/blob/12017a58e8d4e173016af5166a722b4603c41b81/components/gitpod-protocol/go/reconnecting-ws.go#L227)]